### ISSUE TITLE
Explain input and output of reformer

### ISSUE THREAD

**Issue by phongnhhn92** [lucidrains/reformer-pytorch#4]
Hello,
I am trying to understand input and output to your Reformer. I am following your example code
```
import torch
from reformer_pytorch import Reformer

model = Reformer(
    emb = 512,
    depth = 12,
    max_seq_len = 8192,
    num_tokens= 20000,
    heads = 8,
    lsh_dropout = 0.1,
    causal = True,        # auto-regressive or not
    bucket_size = 64,     # average size of qk per bucket, 64 was recommended in paper
    n_hashes = 8,         # should keep at 8 per paper
    ff_chunks = 200,      # number of chunks for feedforward layer
    weight_tie = False,   # tie parameters of each layer for no memory per additional depth
    attn_chunks = 8,        # process lsh attention in chunks, only way for memory to fit when scaling to 16k tokens
    use_full_attn = False   # use full self attention, for comparison
).cuda()

x = torch.randint(0, 20000, (1, 8192)).long().cuda()
y = model(x)
print(x.shape)
print(y.shape)
```
Output:
```
torch.Size([1, 8192])
torch.Size([1, 8192, 20000])
```

So as I understand input x is a 2D tensor with the size of [batch_size,seq_length] and the output y is a 3D tensor with the size of [batch_size, seq_length,_num_tokens]. I wonder why there is a mismatch like this ? 

Comparing to the official Transformer [code](https://pytorch.org/docs/master/_modules/torch/nn/modules/transformer.html#TransformerEncoder.forward), I have a simple example:
```
import torch
from torch.nn.modules.transformer import *
transformer_model = Transformer(nhead=16, num_encoder_layers=12)
src = torch.rand((10, 32, 512))
tgt = torch.rand((20, 32, 512))
out = transformer_model(src, tgt)
print(out.shape)
```
Output:
`torch.Size([20, 32, 512])`

As you can see, the input and output to the Transformer model is a 3D tensor with the size of [seq_length,batch_size,emb_size]. Is there any method I can do the same thing with your Reformer implementation?

**Comment by lucidrains** [comment]
Hi @phongnhhn92 . I think part of the confusion is that the name Transformers is thrown around, and refers to many different things. The Transformer you are importing actually refers to the original Transformer, complete with the encoder and decoder, but also seems to exclude token + positional embedding and the final logits of each token in a full language model.

To keep things consistent, I could rename the current Reformer -> ReformerLM, and make the Reformer behave like the encoder block of the Transformer. I'm afraid Reformer, at the moment, cannot behave like the decoder part yet, that is still a work in progress.

**Comment by lucidrains** [comment]
@phongnhhn92 what are you trying to build?

**Comment by lucidrains** [comment]
@phongnhhn92  i've separated ReformerLM from Reformer. The class Reformer should work as you expect now
